### ISSUE TITLE
Normalize emails before validating uniqueness

### ISSUE THREAD

**Issue by hendi** [digitallyinduced/ihp#164]
The documentation currently shows misleading/dangerous example code for email uniquess:

![image](https://user-images.githubusercontent.com/550218/85844646-5ad9d900-b7a3-11ea-88ae-387bf200c6d3.png)

This misses different case (foo@example.com vs. FOO@example.com) as well as unicode normalization (e.g. `ä` (`U+00E4 latin small letter A with diaeresis`) vs `a:` (`U+0041 latin small letter A` `U+0308 combining diaeresis`)

- [ ] Update documentation of `validateIsUnique` to mention possible pitfalls
- [ ] Provide a `normalizeEmail` function which can be used with `user |> modify #email normalizeEmail |> validateIsUnique #email`

**Comment by kodeFant** [comment]
Maybe a slightly seperate issue, but on usernames it would be nice to be able to case-insensitively validate uniqueness without modifying the actual casing.

**Comment by themmes** [comment]
Also stumbling on this, I doubt that it's just a documentation issue. Because whenever you would normalise in `CreateUserAction` you would also need to apply that same logic to `CreateSessionAction` which is library code, correct?

**Comment by hendi** [comment]
You're right, there are several places where this needs to be handled. For email addresses that are used for login, we should normalize them "everywhere" they are used (i.e. on create, read, update and delete) and store the normalized version in the database to do so efficiently.

Email addresses and usernames should be normalized differently.

**email address**
- unicode normalization (e.g. `ä` (`U+00E4 latin small letter A with diaeresis`) vs `ä` (`U+0041 latin small letter A` `U+0308 combining diaeresis`)
- provider specific normalization (e.g. `mr.x+123@gmail.com` -> `mrx@gmail.com`)
- lower casing (e.g. `MrX@GMail.com` -> `mrx@gmail.com`)
- domain normalization (e.g. `hüächa.de` -> `xn--hcha-loa7l.de`)

**user names**
- unicode normalization
- lower casing can be done during lookup

This issue is about email address normalization, but in implementing this I'll create a (unicode-)normalize function that can be used for user names as well.

**Comment by fegu** [comment]
These are good points. But lowercasing would solve most problems on its own and is far easier to implement correctly. Provider-specific normalization is not necessarily wanted, users sign up with these for a reason.

**Comment by hendi** [comment]
Unicode normalization is authorization credentials is of paramount importance, see for example [Spotify's blog post](https://engineering.atspotify.com/2013/06/18/creative-usernames/) about this. This is a real thread and was used for account hijacking not just in 2013, though I fail to remember which service was affected by this issue just end of last / begining of this year.

You're absolutely right about not storing provider-normalized email addresses in the database! The `+` part is manually set by a user for a reason, we shouldn't remove it. Though on the other hand, we don't want `...+1@gmail.com`, `...+2@gmail.com` and so on each sign up for our free trial. 

Let's start with security first though (i.e. lower case + unicode normalization + punycode normalization) and keep the provider specific one for another issue.
